lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs

```diff
@@ -0,0 +1,801 @@
+                sourceOutputAssets = TokenMap.getAssets output
+                targetOutputAssets = TokenMap.getAssets targetOutput
+
+    safeMerge :: TokenMap -> TokenMap -> Maybe TokenMap
+    safeMerge a b
+        | isSafe = Just value
+        | otherwise = Nothing
+      where
+        isSafe = (&&)
+            (txOutputHasValidSizeIfAdaMaximized constraints value)
+            (txOutputHasValidTokenQuantities constraints value)
+        value = a <> b
+
+--------------------------------------------------------------------------------
+-- Splitting output values
+--------------------------------------------------------------------------------
+
+splitOutputIfLimitsExceeded
+    :: TxSize size
+    => TxConstraints size
+    -> TokenMap
+    -> NonEmpty TokenMap
+splitOutputIfLimitsExceeded constraints =
+    splitOutputIfTokenQuantityExceedsLimit constraints >=>
+    splitOutputIfSizeExceedsLimit constraints
+
+splitOutputIfSizeExceedsLimit
+    :: TxSize size
+    => TxConstraints size
+    -> TokenMap
+    -> NonEmpty TokenMap
+splitOutputIfSizeExceedsLimit constraints value
+    | txOutputHasValidSizeIfAdaMaximized constraints value =
+        pure value
+    | otherwise =
+        split value >>= splitOutputIfSizeExceedsLimit constraints
+    | otherwise =
+        pure value
+  where
+    split = flip TokenMap.equipartitionAssets (() :| [()])
+
+splitOutputIfTokenQuantityExceedsLimit
+    :: TxConstraints size
+    -> TokenMap
+    -> NonEmpty TokenMap
+splitOutputIfTokenQuantityExceedsLimit
+    = flip TokenMap.equipartitionQuantitiesWithUpperBound
+    . txOutputMaximumTokenQuantity
+
+-- | Checks that an output has a valid size even if it is assigned the maximum
+--   possible ada quantity.
+--
+-- Using this function to check all outputs provided to 'balance' will ensure
+-- that it has complete freedom to adjust the ada quantities of those outputs,
+-- without exceeding the output size limit.
+--
+txOutputHasValidSizeIfAdaMaximized
+    :: TxSize size => TxConstraints size -> TokenMap -> Bool
+txOutputHasValidSizeIfAdaMaximized constraints output =
+    txOutputHasValidSize constraints (TokenBundle maxBound output)
+
+--------------------------------------------------------------------------------
+-- Minimizing fees
+--------------------------------------------------------------------------------
+
+minimizeFee
+    :: TxConstraints size
+    -> (Coin, NonEmpty TokenBundle)
+    -- ^ Fee excess and output bundles.
+    -> (Coin, NonEmpty TokenBundle)
+    -- ^ Fee excess and output bundles after optimization.
+minimizeFee constraints (currentFeeExcess, outputs) =
+    NE.fromList <$> run currentFeeExcess (NE.toList outputs) []
+  where
+    run :: Coin -> [TokenBundle] -> [TokenBundle] -> (Coin, [TokenBundle])
+    run (Coin 0) remaining processed =
+        (Coin 0, L.reverse processed <> remaining)
+    run feeExcessRemaining [] processed =
+        (feeExcessRemaining, L.reverse processed)
```

Comment:
Suggestion: remove these reversals, as they are almost certainly unnecessary.

Their existence is a hangover from an earlier iteration of the design, where maintaining the order of outputs was important.

Reply:
Fixed in f9de1b4dad3c642b350217c0afad318ae058e9ec.